azurelinuxagent/ga/agent_update.py

```diff
@@ -0,0 +1,340 @@
+            raise Exception(
+                u"No manifest links found for agent family: {0} for incarnation: {1}, skipping agent update".format(
+                    self._ga_family, self._gs_id))
+        return agent_family_manifests[0]
+
+    @staticmethod
+    def __get_requested_version(agent_family):
+        """
+        Get the requested version from agent family
+        Returns: Requested version if supported and available
+                 None if requested version missing or GA versioning not enabled
+        """
+        if conf.get_enable_ga_versioning() and agent_family.is_requested_version_specified:
+            if agent_family.requested_version is not None:
+                return FlexibleVersion(agent_family.requested_version)
+        return None
+
+    @staticmethod
+    def __get_largest_version(agent_manifest):
+        largest_version = FlexibleVersion("0.0.0.0")
+        for pkg in agent_manifest.pkg_list.versions:
+            pkg_version = FlexibleVersion(pkg.version)
+            if pkg_version > largest_version:
+                largest_version = pkg_version
+        return largest_version
+
+    def __download_and_get_agent(self, goal_state, agent_family, agent_manifest, requested_version):
+        """
+        This function downloads the new agent(requested version) and returns the downloaded version.
+        """
+        if agent_manifest is None:  # Fetch agent manifest if it's not already done
+            agent_manifest = goal_state.fetch_agent_manifest(agent_family.name, agent_family.uris)
+        package_to_download = self.__get_agent_package_to_download(agent_manifest, requested_version)
+        is_fast_track_goal_state = goal_state.extensions_goal_state.source == GoalStateSource.FastTrack
+        agent = GuestAgent.from_agent_package(package_to_download, self._protocol, is_fast_track_goal_state)
+        return agent
+
+    def __get_agent_package_to_download(self, agent_manifest, version):
+        """
+        Returns the package of the given Version found in the manifest. If not found, returns exception
+        """
+        for pkg in agent_manifest.pkg_list.versions:
+            if FlexibleVersion(pkg.version) == version:
+                # Found a matching package, only download that one
+                return pkg
+
+        raise Exception("No matching package found in the agent manifest for requested version: {0} in goal state incarnation: {1}, "
+                        "skipping agent update".format(str(version), self._gs_id))
+
+    @staticmethod
+    def __purge_extra_agents_from_disk(known_agents):
+        """
+        Remove from disk all directories and .zip files of unknown agents
+        (without removing the current, running agent).
+        """
+        path = os.path.join(conf.get_lib_dir(), "{0}-*".format(AGENT_NAME))
+
+        known_versions = [agent.version for agent in known_agents]
+        known_versions.append(CURRENT_VERSION)
```

Comment:
I think you should pass CURRENT_VERSION explicitly as part of known_agents to make the purpose of this method clearer